src/mod/pbkdf2.mod/pbkdf2.c

```diff
@@ -0,0 +1,302 @@
+/* SPDX-License-Identifier: GPL-2.0-only */
+/*
+ * pbkdf2.c -- part of pbkdf2.mod
+ *
+ * Written by thommey and Michael Ortmann
+ *
+ * Copyright (C) 2017 - 2020 Eggheads Development Team
+ */
+
+#include "src/mod/module.h"
+
+#ifdef TLS
+#define MODULE_NAME "encryption2"
+
+#include <resolv.h> /* base64 encode b64_ntop() and base64 decode b64_pton() */
+#include <sys/resource.h>
+#include <openssl/err.h>
+#include <openssl/rand.h>
+
+/* Salt string length */
+#define PBKDF2_SALT_LEN 16 /* DO NOT TOUCH! */
+
+static Function *global = NULL;
+
+/* Cryptographic hash function used. openssl list -digest-algorithms */
+static char pbkdf2_method[28] = "SHA512";
+/* Enable re-encoding of password if pbkdf2-method and / or pbkdf2-rounds
+ * change.
+ */
+static int pbkdf2_re_encode = 0;
+/* Number of rounds (iterations). Default is 5000 like in glibc. */
+static int pbkdf2_rounds = 5000;
+
+static tcl_ints my_tcl_ints[] = {
+  {"pbkdf2-re-encode",  &pbkdf2_re_encode,  0},
+  {"pbkdf2-rounds",     &pbkdf2_rounds,     0},
+  {NULL,                NULL,               0}
+};
+
+static tcl_strings my_tcl_strings[] = {
+  {"pbkdf2-method", pbkdf2_method, 27, 0},
+  {NULL,            NULL,          0,  0}
+};
+
+static char *pbkdf2_close(void)
+{
+  return "You cannot unload the " MODULE_NAME " module.";
+}
+
+static void bufcount(char **buf, int *buflen, int bytes)
+{
+  *buf += bytes;
+  *buflen -= bytes;
+}
+
+static int b64_ntop_without_padding(u_char const *src, size_t srclength,
+                                    char *target, size_t targsize)
+{
+  char *c;
+
+  if (b64_ntop(src, srclength, target, targsize) < 0)
+    return -1;
+  c = strchr(target, '=');
+  if (c) {
+    *c = 0;
+    targsize = (c - target);
+  }
+  return targsize;
+}
+
+/* Return
+ *   hash = "$pbkdf2-<digest>$rounds=<rounds>$<salt>$<hash>" (PHC string format)
+ *     salt and hash = base64
+ *   NULL = error
+ */
+static char *pbkdf2_hash(const char *pass, const char *digest_name,
+                         const unsigned char *salt, unsigned int saltlen,
+                         unsigned int rounds)
+{
+  const EVP_MD *digest;
+  int digestlen, ret;
+  int outlen, restlen;
+  static char *out; /* static object is initialized to zero (Standard C) */
```

Comment:
Can we change this to have the caller pass a dstbuf instead? Re-entrancy probably won't matter but it's an unusual calling convention "result is valid until function is called again". If not, that should at least be boldly documented, because it's new to eggdrop.
If worried about dst buffer size, we can require the dst buffer size being passed as well and being at least the max length of the PHC entry, just document it with a comment

Reply:
yes we could. and usually i would. so le me explain why i did not. i wanted to encypsulate/hide all pbkdf2 logic into the pbkdf2 mod. if the caller were to initialize the buffer, the caller wouldneed more knowledge of pbkdf2 stuff, like:
```

   95   outlen = strlen("$pbkdf2-") + strlen(digest_name) +                            
   96            strlen("$rounds=4294967295$i") + B64_NTOP_CALCULATE_SIZE(saltlen) +   
   97            1 + B64_NTOP_CALCULATE_SIZE(digestlen);
```

now we could

1. export some variables from pbkdf2
or 2. put the needed variables into eggdrop core
or 3. make a new pbkdf2 mod function for getting that outlen
or 4. dont use the exact outlen and let the caller make an oversized array in this case. like char[1024]

so, pls tell me, which path you prefer me to follow now.

BTW: yes, we can boldy document it, but it is _not_ new to eggdrop. example:

https://github.com/eggheads/eggdrop/blob/develop/src/dcc.c#L93

Reply:
oh, and  came up with a 5th solution: do it like snprintf, if you give that function a NULL instead of a buffer, it could return the size, like snprintf() does. so you can call it twice.